servers/src/main/java/tachyon/master/block/BlockMaster.java

```diff
@@ -0,0 +1,571 @@
+
+  // Block metadata management.
+  /**
+   * blocks ever placed in all Tachyon workers, including both the active blocks and lost ones. This
+   * state much be journaled.
+   */
+  private final Map<Long, MasterBlockInfo> mBlocks = new HashMap<Long, MasterBlockInfo>();
+  /**
+   * This state much be journaled.
+   */
+  private final BlockIdGenerator mBlockIdGenerator = new BlockIdGenerator();
+  private final Set<Long> mLostBlocks = new HashSet<Long>();
+  private final BlockingQueue<MasterWorkerInfo> mLostWorkers =
+      new LinkedBlockingQueue<MasterWorkerInfo>();
+  private final TachyonConf mTachyonConf;
+
+  /**
+   * the service that detects lost worker nodes, and tries to restart the failed workers
+   */
+  private Future<?> mLostWorkerDetectionService;
+
+  // Worker metadata management.
+  private final IndexedSet.FieldIndex<MasterWorkerInfo> mIdIndex =
+      new IndexedSet.FieldIndex<MasterWorkerInfo>() {
+        @Override
+        public Object getFieldValue(MasterWorkerInfo o) {
+          return o.getId();
+        }
+      };
+  private final IndexedSet.FieldIndex<MasterWorkerInfo> mAddressIndex =
+      new IndexedSet.FieldIndex<MasterWorkerInfo>() {
+        @Override
+        public Object getFieldValue(MasterWorkerInfo o) {
+          return o.getAddress();
+        }
+      };
+
+  @SuppressWarnings("unchecked")
+  private final IndexedSet<MasterWorkerInfo> mWorkers =
+      new IndexedSet<MasterWorkerInfo>(mIdIndex, mAddressIndex);
+  // This state must be journaled.
+  private final AtomicLong mNextWorkerId = new AtomicLong(1);
+
+  public static String getJournalDirectory(String baseDirectory) {
+    return PathUtils.concatPath(baseDirectory, Constants.BLOCK_MASTER_SERVICE_NAME);
+  }
+
+  public BlockMaster(Journal journal, TachyonConf tachyonConf) {
+    super(journal,
+        Executors.newFixedThreadPool(2, ThreadFactoryUtils.build("block-master-%d", true)));
+    mTachyonConf = tachyonConf;
+  }
+
+  @Override
+  public TProcessor getProcessor() {
+    return new BlockMasterService.Processor<BlockMasterServiceHandler>(
+        new BlockMasterServiceHandler(this));
+  }
+
+  @Override
+  public String getProcessorName() {
+    return Constants.BLOCK_MASTER_SERVICE_NAME;
+  }
+
+  @Override
+  public void processJournalCheckpoint(JournalInputStream inputStream) throws IOException {
+    // clear state before processing checkpoint.
+    mBlocks.clear();
+
+    super.processJournalCheckpoint(inputStream);
+  }
+
+  @Override
+  public void processJournalEntry(JournalEntry entry) throws IOException {
+    if (entry instanceof BlockIdGeneratorEntry) {
+      mBlockIdGenerator.setNextContainerId(((BlockIdGeneratorEntry) entry).getNextContainerId());
+    } else if (entry instanceof WorkerIdGeneratorEntry) {
+      mNextWorkerId.set(((WorkerIdGeneratorEntry) entry).getNextWorkerId());
+    } else if (entry instanceof BlockInfoEntry) {
+      BlockInfoEntry bie = (BlockInfoEntry) entry;
```

Comment:
Minor style: usually we do the acronyms for exceptions, perhaps use the full words here?